concordium-consensus/src/Concordium/KonsensusV1/Consensus/CatchUp.hs

```diff
@@ -13,3 +54,490 @@ module Concordium.KonsensusV1.Consensus.CatchUp where
+    catchUpLastFinalized peerLFBlock = do
+        if blockHeight lfBlock == blockHeight peerLFBlock
+            then do
+                -- We have the same last finalized block, so no need to catch up finalized blocks.
+                catchUpAfterLastFinalized True
+            else do
+                -- Send the finalized blocks from 1 after the peer's last finalized block,
+                -- skipping while the peer claims to have seen them. Note that, since peerLFBlock
+                -- is finalized, but is not the last finalized block, then we have:
+                -- 0 < blockHeight peerLFBlock + 1 <= blockHeight lfBlock
+                sendBlocksFromHeight True (blockHeight peerLFBlock + 1)
+    -- Send our finalized blocks from the given height up to the last finalized block, if they
+    -- are not in 'peerKnownBlocks'.  The boolean flag 'checkKnown' indicates whether we should
+    -- check if the peer already knows the block.  We assume that if the peer does not know a
+    -- block, then it does not know any children of the block.
+    -- INVARIANT: 0 < hgt <= blockHeight lfBlock
+    sendBlocksFromHeight checkKnown hgt
+        | hgt == blockHeight lfBlock =
+            if checkKnown && isPeerKnown (getHash lfBlock)
+                then catchUpAfterLastFinalized True
+                else do
+                    -- By the invariant, it is not possible for 'lfBlock' to be a genesis block.
+                    return
+                        CatchUpPartialResponseBlock
+                            { cuprNextBlock = toSignedBlock (bpBlock lfBlock),
+                              cuprContinue = catchUpAfterLastFinalized False,
+                              cuprFinish = return Absent
+                            }
+        | otherwise =
+            LowLevel.lookupBlockByHeight hgt >>= \case
+                Nothing -> do
+                    -- This should not fail, since the height is below the height of the last
+                    -- finalized block.
+                    error $ "handleCatchUpRequest: missing finalized block at height " ++ show hgt
+                Just fb
+                    | checkKnown && isPeerKnown (getHash fb) -> do
+                        -- If the peer already knows this block, skip it.
+                        sendBlocksFromHeight checkKnown (hgt + 1)
+                    | otherwise -> do
+                        -- Since 0 < hgt, we can be sure that 'fb' is not a genesis block.
+                        return
+                            CatchUpPartialResponseBlock
+                                { cuprNextBlock = toSignedBlock (LowLevel.stbBlock fb),
+                                  cuprContinue = sendBlocksFromHeight False (hgt + 1),
+                                  cuprFinish = return Absent
+                                }
+    catchUpAfterLastFinalized checkKnown = do
+        catchUpBranches checkKnown checkKnown [] (skovData ^. branches)
```

Comment:
Since this function is called without acquiring any locks, it's rather subtle that this will work correctly and send all the blocks, in the case that finalization happens between, say, `sendBlocksFromHeight` is called and then the continuation is called.

That does appear to be the case since the branches are just kept in memory, and the state is pure, not behind mvars or iorefs, but this is at all not clear just from looking at this function. It also crucially relies on `skovData` being passed in at the top-level.

So I would appreciate if a bit more explanation was given on why state changes in between invoking continuations are correctly handled here. Otherwise it's very hard to maintain this in the future.